protocols/fs/src/server.cpp

```
@@ -889,6 +889,51 @@ async::detached handlePassthrough(smarter::shared_ptr<void> file,
 			if(res.error() == Error::wouldBlock) {
 				resp.set_error(managarm::fs::Errors::WOULD_BLOCK);
 
+				auto ser = resp.SerializeAsString();
+				auto [send_resp] = co_await helix_ng::exchangeMsgs(
+					conversation,
+					helix_ng::sendBuffer(ser.data(), ser.size())
+				);
+				HEL_CHECK(send_resp.error());
+			} else if(res.error() == Error::hostUnreachable) {
+				resp.set_error(managarm::fs::Errors::HOST_UNREACHABLE);
+
+				auto ser = resp.SerializeAsString();
+				auto [send_resp] = co_await helix_ng::exchangeMsgs(
+					conversation,
+					helix_ng::sendBuffer(ser.data(), ser.size())
+				);
+				HEL_CHECK(send_resp.error());
+			} else if(res.error() == Error::accessDenied) {
+				resp.set_error(managarm::fs::Errors::ACCESS_DENIED);
+
+				auto ser = resp.SerializeAsString();
+				auto [send_resp] = co_await helix_ng::exchangeMsgs(
+					conversation,
+					helix_ng::sendBuffer(ser.data(), ser.size())
+				);
+				HEL_CHECK(send_resp.error());
+			} else if(res.error() == Error::netUnreachable) {
+				resp.set_error(managarm::fs::Errors::NETWORK_UNREACHABLE);
+
+				auto ser = resp.SerializeAsString();
+				auto [send_resp] = co_await helix_ng::exchangeMsgs(
+					conversation,
+					helix_ng::sendBuffer(ser.data(), ser.size())
+				);
+				HEL_CHECK(send_resp.error());
+			} else if(res.error() == Error::destAddrRequired) {
+				resp.set_error(managarm::fs::Errors::DESTINATION_ADDRESS_REQUIRED);
+
+				auto ser = resp.SerializeAsString();
+				auto [send_resp] = co_await helix_ng::exchangeMsgs(
+					conversation,
+					helix_ng::sendBuffer(ser.data(), ser.size())
+				);
+				HEL_CHECK(send_resp.error());
+			} else if(res.error() == Error::addressNotAvailable) {
+				resp.set_error(managarm::fs::Errors::ADDRESS_NOT_AVAILABLE);
+
```

I would suggest creating a mapping function Error::* -> managarm::fs::Errors::* that would return an optional or a boolean. If mapping is succeesful, error response is set and exchangeMsgs is called avoiding code duplication.

---

:+1: for the mapping function.

But let's make it map one type to the other and panic (= fail an assertion) if the argument is invalid; there is no reasonable way to handle errors inside the mapping function.

---

Just return false or {} in case of mismatching.

---

But what do we send to the client in that case? There is no meaningful way to handle that (if we send a "internal error" message to the client, the client will just crash). Passing an invalid value is a contract violation, not a runtime error.

---

There is already and else handling the mismatch.

```
managarm::fs::Errors resultedError;
if (map(inputError, resultedError)) {
   // set error, exchange msgs
} else {
  // reuse existing "else" code
}
```